workflow/executor/resource.go

```diff
@@ -265,6 +272,12 @@ func (we *WorkflowExecutor) SaveResourceParameters(resourceNamespace string, res
 		if param.ValueFrom == nil {
 			continue
 		}
+		if we.Template.Resource.IgnoreNotFound && resourceName == "" {
+			// Resource wasn't found, so we populate result with sentinel value
+			output := "<NOT_FOUND>"
```

Comment:
Is there a reason you provided a sentinel value? Using `kubectl` directly doesn't provide one and I don't think any user would expect a value to be filled in if a resource doesn't exist.

Considering your use case from your proposal:
```yaml
...
  - name: check-for-existence
    resource:
      action: get
      ignoreNotFound: true
      manifest: |
        apiVersion: v1
        kind: Pod
        metadata:
          name: does-not-exist
          namespace: argo-workflows
    outputs:
      parameters:
      - name: not-found
        valueFrom:
          jsonPath: '{.metadata.name}'
```
A user would have to know what the sentinel value is before checking it against the result of `'{.metadata.name}'` to determine if a resource actually exists or not. It would be more logical to simply check if `'{.metadata.name}'` is empty or not.

Reply:
The reason that there is a sentinel value is because we have three separate possibilities from any jsonpath expression: 1) non-empty value like "foo", 2) empty value like "{.metadata.annotations.imageregistry}" == "", and 3) non-existent entity.

If we don't provide a sentinel value, it's still possible to detect the difference between cases (2) and (3), but it requires the developer to either use two steps (one with a known existing jsonpath like .metadata.name followed by the desired jsonpath) or by including a marker in the jsonpath expression (e.g. '<{.metadata.annotations.imageregistry}>' which can than be compared with "<>" for empty vs "" for non-existent).  Both of these require considerable extra costs to development, either extra runtime costs or string parsing with every result.

Using a sentinel value provides a nice balance that makes usage simple and efficient for most use cases.  In the case where the user happens to be using the exact sentinel value chosen here, they then have to choose a solution like adding markers to the jsonpath.

Adding a custom sentinel value with an additional flag would also solve that problem, but that would be adding more code and configuration for a use case that is unlikely to occur in actual usage.

Reply:
I see your point, thanks for explaining.

@alexec What do you think about this?  I am leaning towards not including it. Even though it will force developers to have to do an extra check, I think it is more intuitive for users.

Reply:
My first thought wast that it should probably be a sentinel error. They are more clearly defined. However, as I understand it, you want to:

1. Ignore "not found" errors.
2. Allow subsequent steps to understand that the resource was not found and act on it.

Am I correct? If that's the case, could we say that .metadata.creationTImestamp is null?

Reply:
@alexec  The main point of this feature is to disassociate the kubectl not-found error from the workflow error domain.  
Right now, the "resource" template subsumes all kubectl errors, which would normally be a useful behavior, but in the case of kubectl, they made a decision to return an error code when looking for a resource that doesn't exist (or for deleting one).

While this can be useful in the guise of a shell script, where you can add per-command error handling, in argo-workflows there is no per-step handling of errors, so gracefully handling this case is problematic.

We could switch the behavior over wholesale, but that might be surprising to existing users.  This PR just adds the option to the user to set the --ignore-not-found flag on get or delete operations.  With this set, deleting a resource becomes idempotent in the case of a race condition (delete if found, ignore if someone else deleted it first), and getting a resource can now also be used to check for the existence of a resource (something that I need enough that I'm already maintaining a local fork to do so).

The added behavior that we're discussing here is specifically about how to return this additional piece of information to the developer in a way that is useful to consume.  We only return a single string, so we have to choose how to encode the new information.

As you've indicated, we could force all developers who are trying to retrieve some field value to use two steps in their workflow (One to check a value like.metadata.creationTimestamp and one to check the desired value).  This performs an additional API RPC call, adds argo executor overhead, and requires additional workflow complexity.

We could also suggest that the user add their own markers, but that requires them to split out the marker on every step to get a useful value.

The code here returns a fixed sentinel value that is unlikely to appear in the wild.  I've chosen the characters to be outside of the set used in kubernetes identifiers, and unlikely to occur in a natural language sentence.  This way, the check is as simple as a string comparison for the sentinel.